### PR TITLE
refactor(ekf_localizer): isolate ekf into independent module

### PR DESCRIPTION
**Pull request by kminoda**
## Description

This PR isolates EKF into a module, independent from the original node implementation. With this PR, EKF node no longer need to handle `ekf_` directly, which is rather a lower level class. Due to this reason, the original code have to handle `ekf_` class multiple times, making the code more difficult to read.

With this PR, EKF is isolated from other features (ROS 2 interfaces, Simple1dFilters, etc). This should drastically improve the readability of the node and easier for other developers to add any features to the node.

## Tests performed

- Tested with sample rosbag on Autoware tutorial and confirmed that the estimation results is overall good
- Confirmed that all the error logs are properly output when necessary as it was originally.

## Effects on system behavior

Not applicable

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: localization/ekf_localizer/include/ekf_localizer/extended_kalman_filter_module.hpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
+#include <geometry_msgs/msg/transform_stamped.hpp>
+#include <geometry_msgs/msg/twist_stamped.hpp>
+#include <geometry_msgs/msg/twist_with_covariance_stamped.hpp>
+
+#include <memory>
+
+struct EKFDiagnosticInfo
```
Comment (YamatoAndo) [inline localization/ekf_localizer/include/ekf_localizer/extended_kalman_filter_module.hpp:34]: build error

> /__w/autoware.universe/autoware.universe/localization/ekf_localizer/include/ekf_localizer/ekf_localizer.hpp:51:8: error: redefinition of ‘struct EKFDiagnosticInfo’
>    51 | struct EKFDiagnosticInfo
>       |        ^~~~~~~~~~~~~~~~~
> In file included from /__w/autoware.universe/autoware.universe/localization/ekf_localizer/include/ekf_localizer/ekf_localizer.hpp:19,
>                  from /__w/autoware.universe/autoware.universe/localization/ekf_localizer/src/ekf_localizer.cpp:15:
> /__w/autoware.universe/autoware.universe/localization/ekf_localizer/include/ekf_localizer/extended_kalman_filter_module.hpp:34:8: note: previous definition of ‘struct EKFDiagnosticInfo’
>    34 | struct EKFDiagnosticInfo

Comment (kminoda) [reply localization/ekf_localizer/include/ekf_localizer/extended_kalman_filter_module.hpp:34]: Thank you! Fixed: https://github.com/autowarefoundation/autoware.universe/pull/5424/commits/41da3836a33e4897a91ff1f6e213704f5a83a48f

File: localization/ekf_localizer/src/ekf_localizer.cpp [hunk 2]
```diff
@@ -190,12 +160,18 @@ void EKFLocalizer::timerCallback()
     stop_watch_.tic();
 
     // save the initial size because the queue size can change in the loop
-    const size_t n = pose_queue_.size();
-    for (size_t i = 0; i < n; ++i) {
+    const auto t_curr = this->now();
+    for (size_t i = 0; i < pose_queue_.size(); ++i) {
```
Comment (KYabuuchi) [inline localization/ekf_localizer/src/ekf_localizer.cpp:164]: `pose_queue_.size()` might decrease within the loop, so it should store the size as `n` beforehand and then loop `n` times.

File: localization/ekf_localizer/src/ekf_localizer.cpp [hunk 3]
```diff
@@ -219,10 +194,11 @@ void EKFLocalizer::timerCallback()
     stop_watch_.tic();
 
     // save the initial size because the queue size can change in the loop
-    const size_t n = twist_queue_.size();
-    for (size_t i = 0; i < n; ++i) {
+    const auto t_curr = this->now();
+    for (size_t i = 0; i < twist_queue_.size(); ++i) {
```
Comment (KYabuuchi) [inline localization/ekf_localizer/src/ekf_localizer.cpp:198]: `twist_queue_.size()` might decrease within the loop, so it should store the size as `n` beforehand and then loop `n` times.

File: localization/ekf_localizer/src/extended_kalman_filter_module.cpp [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+    P(IDX::YAWB, IDX::YAWB) = 50.0;  // for yaw bias
+  }
+  P(IDX::VX, IDX::VX) = 1000.0;  // for vx
+  P(IDX::WZ, IDX::WZ) = 50.0;    // for wz
+
+  kalman_filter_.init(X, P, params_.extend_state_step);
+}
+
+void ExtendedKalmanFilterModule::initialize(
+  PoseWithCovariance & initial_pose, geometry_msgs::msg::TransformStamped & transform)
```
Comment (KYabuuchi) [inline localization/ekf_localizer/src/extended_kalman_filter_module.cpp:57]: You can make it const argument. (Of course, hpp needs to be changed as well.)
```suggestion
void ExtendedKalmanFilterModule::initialize(
  const PoseWithCovariance & initial_pose, const geometry_msgs::msg::TransformStamped & transform)
```

File: localization/ekf_localizer/include/ekf_localizer/extended_kalman_filter_module.hpp
Comment (KYabuuchi) [inline localization/ekf_localizer/include/ekf_localizer/extended_kalman_filter_module.hpp:1]: Is there a specific reason for naming as `ExtendedKalmanFilterModule(extended_kalman_filter_module.hpp)` In this package, we usually use names like `EKFHoge(ekf_hoge)`.

File: localization/ekf_localizer/src/ekf_localizer.cpp (hunk 2 above)
Comment (kminoda) [reply localization/ekf_localizer/src/ekf_localizer.cpp]: Tashikani!!!!
Thanks :bow: Fixed here: https://github.com/autowarefoundation/autoware.universe/pull/5424/commits/048faf9e4767469a200a71d37957dcd39469191b

File: localization/ekf_localizer/src/ekf_localizer.cpp (hunk 3 above)
Comment (kminoda) [reply localization/ekf_localizer/src/ekf_localizer.cpp]: Fixed here: https://github.com/autowarefoundation/autoware.universe/pull/5424/commits/048faf9e4767469a200a71d37957dcd39469191b

File: localization/ekf_localizer/src/extended_kalman_filter_module.cpp (hunk 4 above)
Comment (kminoda) [reply localization/ekf_localizer/src/extended_kalman_filter_module.cpp]: Fixed here: https://github.com/autowarefoundation/autoware.universe/pull/5424/commits/048faf9e4767469a200a71d37957dcd39469191b

File: localization/ekf_localizer/include/ekf_localizer/extended_kalman_filter_module.hpp
Comment (kminoda) [reply localization/ekf_localizer/include/ekf_localizer/extended_kalman_filter_module.hpp:1]: Not specifically, and true that it's better to follow the convention as it is understandable and yet making the file name short enough. Fixed here: https://github.com/autowarefoundation/autoware.universe/pull/5424/commits/048faf9e4767469a200a71d37957dcd39469191b

**Review by KYabuuchi — APPROVED** [review]
LGTM (I am not a code owner though)

**Review by YamatoAndo — APPROVED** [review]

**Pull request merged**